### DISCUSSION TITLE
Device Module Tabs not showing

### DISCUSSION THREAD

**Discussion by BriacD** [netbox-community/netbox#14975] [Q&A] [answered]
After using a custom script found at: https://github.com/netbox-community/customizations/blob/master/scripts/add_device_type_components.py

I noticed that even thought the components are added, the tab does not show up.

![image](https://github.com/netbox-community/netbox/assets/10373814/39c1e684-b5e1-4b60-a95d-936e2081b7cf)

It's only reachable via the URL: dcim/devices/12522/power-outlets/ but they do exist.
I can also connect them with cable objects, so Netbox knows they are there but the tab is missing.

I don't believe this a fault with the script but I could be wrong.

Any idea why the webpage is not rendering this element?

**Comment by candlerb** [comment]
Netbox version?

There is a problem with demo.netbox.dev that the cached *counts* of components on a device are not being set when the sample data is imported - and if the cached count is zero, no tab is shown. For example, https://demo.netbox.dev/dcim/devices/88/front-ports/ shows frontports but https://demo.netbox.dev/dcim/devices/88/ has no "front ports" tab.

I don't think anyone got to the bottom of this, but it's possible that custom scripts are affected by the same issue.

Try `/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py calculate_cached_counts` and see if that fixes the tabs. If it does, then you can raise this as a reproducible issue. I guess it's something to do with the wacky way that signals and hooks are wired up to trigger magical counter updates.

EDIT: that script calls `bulk_create()`, and the Django [documentation](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#bulk-create) says:

> * The model’s save() method will not be called, and the pre_save and post_save signals will not be sent.

And it looks like `post_save()` is [used](https://github.com/netbox-community/netbox/blob/v3.7.1/netbox/utilities/counters.py#L48-L62) to update the counters.

**Reply by candlerb** [reply, marked as answer]
Try this fix: https://github.com/netbox-community/customizations/pull/101

**Reply by BriacD** [reply]
Thanks for the quick reply and breakdown of the issue.
That seems to have did the trick. I was able to update my PDUs and outlets are now showing. I will test with other components to be sure.
